### ISSUE TITLE
New module "ParallelText"

### ISSUE THREAD

**Issue by cysouw** [cldf/cldf#50]
**Proposal**

I would like to propose a new module `ParallelText` for the CLDF, which uses two components: *forms* (i.e. `FormTable`) and a new component for *functional equivalents* (`FunctionalEquivalentsTable`).

The structure of parallel texts are very similar to wordlists in that there are parallel instances ("translations") for the same content, so they basically can be encoded using a [FormTable](https://github.com/cysouw/cldf/tree/master/components/forms).

The analysis of a parallel text is somewhat similar to the analysis of cognates, in that parts of the strings are to be aligned. However, these alignments are not cognates, but functional equivalents. Also, the alignments are mostly not nicely linear as with sound alignments, so it is mostly not possible to solve the multialignment by inserting gaps (there is also a lot of many-to-many alignments). Because of this I would like to propose a new component, the `FunctionalEquivalentsTable` (any improvements on this cumbersome name are welcome!).

The `FunctionalEquivalentsTable` encodes the parts from the different parallel texts that are functionally equivalent. In its most general form, this table is a list of counts/slices from the parallel texts (separated by spaces) that are aligned into `FunctionalEquivalentSets`.

```
{
    "url": "functionalEquivalents.csv", 
    "dc:conformsTo": "TODO", 
    "tableSchema": {
        "columns": [
            {
                "name": "ID", 
                "required": true, 
                "propertyUrl": "http://cldf.clld.org/v1.0/terms.rdf#id", 
                "datatype": {
                    "base": "string", 
                    "format": "[a-zA-Z0-9_\\-]+"
                }
            }, 
            {
                "name": "Form_ID",
                "required": true, 
                "propertyUrl": "http://cldf.clld.org/v1.0/terms.rdf#formReference", 
                "datatype": "string"
            }, 
            {
                "name": "Slice/Count", 
                "required": true, 
                "propertyUrl": "http://cldf.clld.org/v1.0/terms.rdf#slice", 
                "datatype": {
                    "base": "string", 
                    "format": "\\d+(:\\d+)?"
                }, 
                "separator": ","
            }, 
            {
                "name": "FunctionalEquivalentSet", 
                "required": true, 
                "propertyUrl": "TODO", 
                "datatype": "string"
            }
        ]
    }
}
```

See for example [this](https://github.com/cysouw/multialignment-of-paralleltext) multialignment of Germanic bibles (Note that the format here is no yet in accordance with the CLDF). Also nicely illustrated there is the non-linearity of the alignments, e.g. in [this visualisation](https://github.com/cysouw/multialignment-of-paralleltext/blob/master/visualisation/40002001.pdf).

**Comment by cysouw** [comment]
**Question**

In case of parallel texts from highly similar languages (e.g. dialects) it turns out that multi-alignment is possible by inserting gaps, much like sound alignments for cognates. See for example [this](https://github.com/cysouw/wenker/blob/master/data/WS9/WS9_1syntax.txt) case of dialect mutli-alignment.

The question now is whether we could allow two different ways to encode a particular table. The most general format for the `FunctionalEquivalensTable` is the one given above (with sets of slices). However, for dialects it might be more practical to use a format as given below (with space-delimited alignments). Would it lead to problems when we allow two different formats for the same component? Or can we just treat both formats as alternative options within one specification?

```
{
    "url": "functionalEquivalents.csv", 
    "dc:conformsTo": "TODO", 
    "tableSchema": {
        "columns": [
            {
                "name": "ID", 
                "required": true, 
                "propertyUrl": "http://cldf.clld.org/v1.0/terms.rdf#id", 
                "datatype": {
                    "base": "string", 
                    "format": "[a-zA-Z0-9_\\-]+"
                }
            }, 
            {
                "name": "Form_ID", 
                "required": true, 
                "propertyUrl": "http://cldf.clld.org/v1.0/terms.rdf#formReference", 
                "datatype": "string"
            }, 
            {
                "name": "Alignment", 
                "required": false, 
                "propertyUrl": "http://cldf.clld.org/v1.0/terms.rdf#alignment", 
                "datatype": "string", 
                "separator": " "
            }
        ]
    }
}
```

**Comment by cysouw** [comment]
**Question**

If we allow for alternative encodings of the same table, then the components `cognates` and `partialcognates` can be merged into one component, just allowing optional slices. I think that this would be a good idea, but maybe I'm missing something?

**Comment by LinguList** [comment]
We had this discussion about partial cognates and cognates in the past, and I would basically agree with @cysouw, that partialcognates, i.e., the use of slices, is the general case that includes the normal cognates (in fact, I now even think that people should never do normal cognates, as most data is partial in some sense). We left the two types, since the majority of data we encounter has "normal cognates", so partial cognates are a very rare case for the time being. There is also an advantage to distinguish partial and normal cognates, as people use partial cognate insights when deciding on binary cognacy. We recently tested a consistent way of doing so, as most algorithms still only allow normal cognates (phylogeny stuff). If we allow for several alignments and cognate sets of the same forms.csv, it is of course clear, that we could just add two versions, one with and one without slices. But the question is also, whether we would not like to keep the distinction between partial and "normal" cognates for reasons of transparency, since handling normal cognates is algorithmically much less complex, and the majority of data is still, unfortunately, encoded in this way.

**Comment by xrotwang** [comment]
I wouldn't necessarily create a new module for this, but a new component `FunctionalEquivalensTable` sounds like a really good addition. I even like the name! Explicit is better than implicit, after all.

Regarding alternative encodings of special cases: I agree that cognates and partialcognates should be merged. Whether one is dealing with non-partial cognates is already a property of the data after all, so it shouldn't have an effect in the schema, i.e. the metadata.

By the same token I'd argue that the question whether one is dealing with dialectal variation in parallel texts, should be answered by the data and not the schema.

**Comment by cysouw** [comment]
The component `functionalEquivalentsTable` (with a 't' in the middle) is of course a very general component, that can be used more widely.

The module `parallelText` was just inspired by the other modules: I though that such a module is a kind of container for a collection of components with a particular semantics. In this way, parallel texts are a new kind of comparative data that is not yet part of CLDF, but I would like to add it.

Or am I misunderstanding the role of `modules`?

**Comment by cysouw** [comment]
@xrotwang `"By the same token I'd argue that the question whether one is dealing with dialectal variation in parallel texts, should be answered by the data and not the schema."`

The problem is not the kind of data, that was just an example. The problem is that there are two ways that I would like to use  the `functionalEquivalentsTable`:

- either by explicit reference ("this part of string X belongs to alignment A1")
- or by using secondary separators ("string X consists of five alignments with parts 'a - b c d')

Basically, the second approach is a special case of the first approach, but with some data it is much easier to handle in the second approach. So, can we define a table with two different kinds of encodings? If yes, then I'll happy :-)

**Comment by xrotwang** [comment]
When it comes to implementing support in tools stuff like "a table with two different kinds of encodings" typically comes back to haunt you :)

But in this particular case it seems as if just adding the `Alignment` column to your first table description and making `Alignment` and `Slices` optional, you could have both. It would still be a bit of a hassle for tools to compute the missing column from the one given, but I guess that may be ok.

Still, considering that such datasets will typically be assembled by software, wouldn't it be better for CLDF consumers to define just one variant and have the data-creation software deal with the problem of converting between the equivalent descriptions?

**Comment by LinguList** [comment]
Short question to @xrotwang, if we delete the partial cognates and make cognates and alignments more generic, to account for both, depending on the presence of slices, it would be interesting to have some solution how to deal with different cognate sets in the same datasets (as this occurs in those cases where I start doing partial cognates and then use our new guidelines to binarise them). 

As far as I understand the current situation, I could just add different IDs in cognates.csv, which describe the same word. This would essentially also mean that I could have both partial cognates and normal cognates, but even different experts in the same cognates.csv, right? For export to cldf from lingpy, we'd need to modify the code, but our current spec is not against that, so a file could look like:

```csv
ID,Form_ID,Cognateset_ID,Alignment,Source,Slice
cog-1,word-1,cog-1,b l a + b l a,List2017,
cog-2,word-2,cog-1,b l - - - - -,List2017,
part-1,word-1,cog-1,b l a,List2017a,0 1
part-2,word-2,cog-1,b l -,List2017a,0 1
part-3,word-3,cog-1,b l a,List2017a,1 2
```

In current LingPy export, ID and Form_ID are identical, but this does not necessarily need to be the case, as we'd need many IDs for the same form anyway if we have partial cognates.
